tests/python/contrib/test_hexagon/test_conv2d_conv2d.md

```diff
@@ -0,0 +1,860 @@
+
+## Constants
+
+| Constant           | Value |
+| ------------------ | ----- |
+| Conv2d #2 Pad      | 0     |
+| Conv2d #1 Dilation | 1     |
+| Conv2d #2 Dilation | 1     |
+
+## Shapes and Layouts
+
+The input is provided and padded in logical layout and then packed into its physical layout prior to compute.  Logical layout / shape information is provided as a reference for phsyical tensors.
+
+| Tensor       | Type     | Layout      | Shape                  | Logical Layout | Logical Shape    |
+| ------------ | -------- | ----------- | ---------------------- | -------------- | ---------------- |
+| Input        | Logical  | NHWC        | [1, 64, 64, 128]       |                |                  |
+| Padded Input | Logical  | NHWC        | [1, 64, 64, 128]       |                |                  |
+| Packed Input | Physical | NHWC8h8w32c | [1, 8, 8, 4, 8, 8, 32] | NHWC           | [1, 64, 64, 128] |
+| Filter 1     | Physical | OIHW8i32o4i | [4, 4, 3, 3, 8, 32, 4] | OIHW           | [128, 128, 3, 3] |
+| Temp Output  | Physical | NHWC8h8w32c | [1, 8, 8, 4, 8, 8, 32] | NHWC           | [1, 62, 62, 128] |
+| Filter 2     | Physical | OIHW8i32o4i | [4, 4, 3, 3, 8, 32, 4] | OIHW           | [128, 128, 3, 3] |
+| Output       | Physical | NHWC8h8w32c | [1, 8, 8, 4, 8, 8, 32] | NHWC           | [1, 60, 60, 128] |
+
+## Schedule
+
+This is the conv2d compute schedule:
+
+```
+  for (ko.outer: int32, 0, 2) {
+    for (ho.outer: int32, 0, 4) {
+      // caches computed here
+      for (ko.c.inner: int32, 0, 2) {
+        for (ho.c.inner: int32, 0, 2) {
+          for (wo.c: int32, 0, 8) {
+            for (rc.outer_1: int32, 0, 4) {
+              for (hi.c: int32, 0, 8) {
+                for (wi.c: int32, 0, 8) {
+                  for (rh_1: int32, 0, 3) {
+                    for (rw_1: int32, 0, 3) {
+                      for (ki.c: int32, 0, 32) {
+                        for (rc.inner_1: int32, 0, 32) {
+
+```
+
+The major change here is the presence of the the kernel height `rh` and width `rw` dimensions.  
+
+```
+                  for (rh_1: int32, 0, 3) {
+                    for (rw_1: int32, 0, 3) {
+```
+
+(Same as above) Note that conv2d #1 has an independent loop over the channel out `ko.outer` dimension.  This is because the output channels of conv2d #1 are the input channels to conv2d #2 and we compute over all input channels for each conv2d so we must compute over all output channels of conv2d #1 before we compute conv2d #2.
+
+```
+      for (ko.outer_1: int32, 0, 2) {
+```
+
+(Different from above) Note that conv2d #1 also has an independent loop over some portion of the  `ho.outer` dimension.  This is due to the fact that the 3x3 filter will "fall off the bottome" of the input and thus the vertically adjacent "full width" and "full depth" slice of the input must be a) prefetched into the input cache for conv2d #1 and b) produced in the temporary output cache of conv2d #2.
+
+```
+        for (ho.outer_1: int32, 0, 2) {
```

Comment:
So this is (re)computing every horizontal slice of the first conv2d twice?

Reply:
Added some more info in the README to help explain.  Short story:  the first conv2d must produce sufficient output in the height dimension prior to the second conv2d starting.

Reply:
To be clear, this implies that when striping full width slices along h, the compute will look like,

conv2d_0_hslice_0
**conv2d_0_hslice_1** 
conv2d_1_hslice0

**conv2d_0_hslice1**
conv2d_0_hslice2
conv2d_1_hslice1

conv2d_0_hslice2
conv2d_0_hslice3
conv2d_1_hslice2

Which effectively means we do 2x the amount of compute for conv2d_0 relative to the non-striping case?